sqlmesh/core/context.py

```diff
@@ -1597,53 +1599,156 @@ def table_diff(
             temp_schema: The schema to use for temporary tables.
 
         Returns:
-            The TableDiff object containing schema and summary differences.
+            The list of TableDiff objects containing schema and summary differences.
         """
-        source_alias, target_alias = source, target
 
-        adapter = self.engine_adapter
+        table_diffs: t.List[TableDiff] = []
 
-        if model_or_snapshot:
-            model = self.get_model(model_or_snapshot, raise_if_missing=True)
-            adapter = self._get_engine_adapter(model.gateway)
+        # Diffs multiple or a single model across two environments
+        if model_or_snapshot or select_model:
             source_env = self.state_reader.get_environment(source)
             target_env = self.state_reader.get_environment(target)
-
             if not source_env:
                 raise SQLMeshError(f"Could not find environment '{source}'")
             if not target_env:
-                raise SQLMeshError(f"Could not find environment '{target}')")
-
-            # Compare the virtual layer instead of the physical layer because the virtual layer is guaranteed to point
-            # to the correct/active snapshot for the model in the specified environment, taking into account things like dev previews
-            source = next(
-                snapshot for snapshot in source_env.snapshots if snapshot.name == model.fqn
-            ).qualified_view_name.for_environment(source_env.naming_info, adapter.dialect)
-
-            target = next(
-                snapshot for snapshot in target_env.snapshots if snapshot.name == model.fqn
-            ).qualified_view_name.for_environment(target_env.naming_info, adapter.dialect)
-
-            source_alias = source_env.name
-            target_alias = target_env.name
-
-            if not on:
-                on = []
-                for expr in [ref.expression for ref in model.all_references if ref.unique]:
-                    if isinstance(expr, exp.Tuple):
-                        on.extend(
-                            [key.this.sql(dialect=adapter.dialect) for key in expr.expressions]
-                        )
-                    else:
-                        # Handle a single Column or Paren expression
-                        on.append(expr.this.sql(dialect=adapter.dialect))
+                raise SQLMeshError(f"Could not find environment '{target}'")
+
+            modified_snapshots: t.Set[ModelOrSnapshot] = (
+                {model_or_snapshot} if model_or_snapshot else set()
+            )
+            if select_model:
+                models_to_diff = self._new_selector().expand_model_selections(select_model)
+                target_snapshots = {
```

Comment:
For models that were excluded this way we need some sort of message indicating that the same model version is used in both environments and there's no point in diffing. Otherwise, it will be confusing to the user why the models they selected haven't been compared.